common/workunit/workunit.cpp

```diff
@@ -6863,40 +6834,51 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
     throwUnexpected();
 }
 
-bool extractFromWorkunitDAToken(const char * distributedAccessToken, StringBuffer * wuid, StringBuffer * user, StringBuffer * privKey)
+//Does given token appear to be in the right format. KeyFile and signature are optional
+//    HPCC[u=user,w=wuid]keyFile;signature
+bool isWorkunitDAToken(const char * distributedAccessToken)
 {
-    if (strncmp(distributedAccessToken, "HPCC[", 5))
+    if (strncmp(distributedAccessToken, "HPCC[", 5) ||
+        !strstr(distributedAccessToken+5, "u=") ||
+        !strchr(distributedAccessToken+5, ',')  ||
+        !strstr(distributedAccessToken+5, "w=") ||
+        !strchr(distributedAccessToken+5, ']')  ||
+        !strchr(distributedAccessToken+5, ';'))
     {
-        DBGLOG("Not a valid workunit distributed access token");
         return false;
     }
+    return true;//appears to be a workunit token
+}
 
-    const char * finger;
-    if (user)
+bool extractFromWorkunitDAToken(const char * distributedAccessToken, StringBuffer * wuid, StringBuffer * user, StringBuffer * privKey)
+{
+    if (!isWorkunitDAToken(distributedAccessToken))
     {
-        finger = strstr(distributedAccessToken+5, "u=");
-        if (finger)
-        {
-            finger += 2;
-            while (*finger && *finger != ',' && *finger != ']')
-                user->append(1, (const char *)finger++);
-        }
+        DBGLOG("Not a valid workunit distributed access token");
+        return false;
     }
 
-    if (wuid)
+    //Isolate string between []
+    StringBuffer tokenValues(distributedAccessToken + 5);
+    tokenValues.replace(']', (char)0);
+
+    StringArray nameValues;
+    nameValues.appendList(tokenValues.str(), ",",true);
+
+    if (user || wuid)
     {
-        finger = strstr(distributedAccessToken+5, "w=");
-        if (finger)
+        for (int x = 0; x < nameValues.ordinality(); x++)
         {
-            finger += 2;
-            while (*finger && *finger != ',' && *finger != ']')
-                wuid->append(1, (const char *)finger++);
+            if (user && 0==strncmp(nameValues[x],"u=",2))//Extract user
+                user->append(nameValues[x] + 2);
+            else if (wuid && 0==strncmp(nameValues[x],"w=",2))//Extract wuid
+                wuid->append(nameValues[x] + 2);
         }
     }
 
     if (privKey)
     {
-        finger = strstr(distributedAccessToken+5, "]");
+        const char * finger = strchr(distributedAccessToken+5, ']');
```

Comment:
if you calculate the posistion of the ] above to create the exact content for tokenValues above, then you can just use the saved off position here.